### DISCUSSION TITLE
Restrict private data to the owner user only

### DISCUSSION THREAD

**Discussion by nassim-kachroud** [MichalLytek/typegraphql-prisma#82] [Q&A]
Hey guys!

First, thanks a lot for the efforts in it, it works like a charm!
I'm currently working on a e-commerce GraphQL API and I would like the best solution to protect fetching  private data owned by the logged in user.
I used a auth-checker to make sure that a user is logged in but for example, how can I make sure that users can fetch only orders, coupons... they own?

Thanks a lot!

**Comment by MichalLytek** [comment]
> orders, coupons... they own

It's a business logic so you need to write your own resolvers which will translate `orders` query into proper `where` conditions of prisma client `findMany` method.

**Reply by nassim-kachroud** [reply]
Thanks for your reply! 
I already started working on custom resolvers like this, but as the ExtendedUsers is using Users, I'm forced to add the UsersCrudResolver in the public reservers, so useless at the end.
What is the best solution to only expose UsersExtendedResolver? Set the field "user" of ExtendedUsers different of the Users type?
Thanks again!
![image](https://user-images.githubusercontent.com/17111789/111756859-f9aae600-889a-11eb-90f1-9b96405fa8e4.png)
![image](https://user-images.githubusercontent.com/17111789/111757058-34148300-889b-11eb-97f7-9349b140eaaa.png)

**Reply by MichalLytek** [reply]
Please read in docs how to expose selected crud actions instead of a full CRUD resolver

**Reply by nassim-kachroud** [reply]
Yes I did and it's there that I stuck - taking the same example just above - on register I would expect to just need to add the CreateUsersResolver like this:
![image](https://user-images.githubusercontent.com/17111789/111763278-38906a00-88a2-11eb-8eb1-bbdcf44203d7.png)

**Reply by MichalLytek** [reply]
Every GraphQL schema needs to have at least one query operation. Currently you have only two mutations.

Is this your first project using GraphQL and TypeGraphQL? If so, please try with something simpler that advanced `typegraphql-prisma` integration.

**Reply by nassim-kachroud** [reply]
Yes this is my first big project with TypeGraphQL, learning days and nights on that.
All these automated stuffs lost me a bit ^^
Changing the Login type from mutation to query fixed the problem thanks!
Last point please is about: applyResolversEnhanceMap. It doesn't seem to be taken in account when I write:
```
applyResolversEnhanceMap({
  Users: {
    findUniqueUsers: [Authorized()],
    findFirstUsers: [Authorized()],
    findManyUsers: [Authorized()],
  },
})
```
And then calling ` prisma.users.findUnique()`, I would expect passing through my auth-checker, do I miss something?
![image](https://user-images.githubusercontent.com/17111789/111766611-054fda00-88a6-11eb-8428-68f04d96cf43.png)

Thanks!

**Reply by MichalLytek** [reply]
It's only for the generated resolvers - you can put `@Authorized()` on your custom ones easily 😛

**Reply by nassim-kachroud** [reply]
I mean the findUnique from users was automatically auto-generated. I want to restrict the GetXXX(id) to users that own it.
At the end I think I have a lot of custom things to code as all the relations (addresses, invoices, orders...) that are linked to Users don't have to be accessible if they aren't linked to the logged in user 😛

**Reply by MichalLytek** [reply]
`typegraphql-prisma` is not a Hasura competitor. You can use the generated resolvers for quick prototyping or use the generated types to add custom business logic to your GraphQL API.
Don't expect a ready-for production tool to express your domain in 5 mins.

**Reply by nassim-kachroud** [reply]
Yes I get it and it's really powerful all the generated stuff, really good job mate!

**Reply by nassim-kachroud** [reply]
Hi @MichalLytek!
Why using the **all** throws an error: `[ERROR] 17:32:16 TypeError: Cannot read property 'prototype' of undefined` whereas by specifying the specific method it works?
![image](https://user-images.githubusercontent.com/17111789/112182669-c7ccb300-8bfd-11eb-9c8a-71822f21d3d8.png)
![image](https://user-images.githubusercontent.com/17111789/112182910-019db980-8bfe-11eb-81ed-5447a599e37d.png)

Thanks!

**Reply by MichalLytek** [reply]
Shouldn't that be `_all`? Have you modified the generated code?

**Reply by MichalLytek** [reply]
I have such error for `all`:
```
Type '{ all: never[]; categories: MethodAndPropDecorator[]; }' is not assignable to type 'Partial<Record<"_all" | keyof CategoryCrudResolver, MethodDecorator[]>>'.
```

**Reply by nassim-kachroud** [reply]
Same with **_all**
![image](https://user-images.githubusercontent.com/17111789/112185046-23983b80-8c00-11eb-9719-38532c3c5cf8.png)

**Reply by MichalLytek** [reply]
Please create a repository with a minimal reproducible code example. Which means schema.prisma + the index file, not a dump of 100 resolvers you have in your project.

**Reply by nassim-kachroud** [reply]
I just did it with only 1 table Categories: same error...
![image](https://user-images.githubusercontent.com/17111789/112187683-9bfffc00-8c02-11eb-8985-59333d5fbff1.png)

**Reply by nassim-kachroud** [reply]
I uploaded the sample code that you can clone from here:

https://github.com/nassim-kachroud/test

**Reply by MichalLytek** [reply]
@nassim-kachroud Do you know that when you read release notes for the newest release, you also need to install it?
Your package.json: `"typegraphql-prisma": "^0.11.2",`
0.12.0 release notes: https://github.com/MichalLytek/typegraphql-prisma/releases/tag/v0.12.0

I won't even address compilation errors caused by wrong imports:
![image](https://user-images.githubusercontent.com/10618781/112721309-78d79400-8f03-11eb-8df2-9e7b6340cf90.png)

**Reply by MichalLytek** [reply]
You should stop using `--transpile-only` and start checking the TypeScript output e.g. via IDE.

**Reply by MichalLytek** [reply]
Upgrading to Prisma 2.19 and this library to 0.12.0 fixes the issue:
```
[INFO] 13:53:55 ts-node-dev ver. 1.1.6 (using ts-node ver. 9.1.1, typescript ver. 4.2.3)
🚀 Server ready at: http://localhost:4444
Query "findFirstCategories" accessed
```
